features/294-login-without-specifying-login-id-key/api.md

```diff
@@ -0,0 +1,104 @@
+For supporting login without specifying `loginID`, have two proposals:
+
+## Proposal 1 - auth gear find corresponding user object automatically
+
+Keeping most of current implementation, auth gear finds possible user object automatically. Consider following use case:
+
+```javascript
+skygear.auth.signup({
+  email: 'example@example.com', 
+  username: 'example'
+}, 'password');
+
+// new:
+skygear.auth.login('example@example.com', 'password')
+.then((user) => {
+  console.log('login successfully');
+});
+```
+
+
+### Possible Problem
+
+If user signup with incorrect `loginIDMetadataKeys`, it may cause user identify miss assign. Consider following scenario:
+
+```javascript
+
+// signup user A
+skygear.auth.signup({
+  email: 'example@example.com', 
+  username: 'example'
+}, 'password');
+
+
+// signup user B
+// B's username is A's email
+skygear.auth.signup({
+  username: 'example@example.com', 
+  email: 'example'
+}, 'password');
+
+// doesn't know which user to match
+// user A: {email: 'example@example.com'}
+// user B: {username: 'example@example.com'}
+skygear.auth.login('example@example.com', 'password');
+```
+
+### Code change
+
+Just brute force and find all possible principals. 
+
+When auth gear is got a login request without specifying `loginIDMetadataKey`, auth gear will try to find all possible principals based on current `loginIDMetadataKeys` setting. 
+
+For example, if a user login with string `"example"`, and `loginIDMetadataKeys=[["username"], ["email"], ["phone"]]`, auth principal store finds all principals which has `{username: "example"}`, `{email: "example"}` and `{phone: "example"}` in `authdata` field.
+
+And then login handler will check input password with stored password, if **only one** of the principals matched, then the user login successfully. If there are more than one principals matched, auth gear will return `Ambiguous loginID` error to the client. 
+
+The major drawback of this solution is the API UX is not good, basically just throw the problem back to the app developer. But the situation of two users uses same password with interchanged `loginID` should be a rare case.
```

Comment:
Random idea: why can't we still check uniqueness for LoginIDs with proposal 1?

Assume the system have `LOGIN_ID_METADATA_KEYS=['email', 'username', ['role', username']]`

Let's say we have a table of loginIDs, and run the following APIs:

```
skygear.auth.signupWithEmail("example@example.com", 'pw');
skygear.auth.signupWithUsername("example@example.com", 'pw'); // UNIQUE LOGINIDS ERROR
skygear.auth.signup({role: 'doctor', username: 'example@example.com'}, 'pw'); // THIS IS OK
```

I think this actually make sense?

Reply:
I consider this yesterday too, but I am a little worried the case is not easy to handle in the client. For example, when user signup, the user input a string as `username`, and the system responses him the error message: `username is already taken as email already`. Of course we can omit the problem though.